packager/media/codecs/ac4_parser.cc

```diff
@@ -0,0 +1,901 @@
+          if (ac4_presentation_v1_info.n_substream_groups == 5) {
+            ac4_presentation_v1_info.n_substream_groups +=
+                ReadAc4VariableBits(reader, 2);
+          }
+          for (int sg = 0; sg < ac4_presentation_v1_info.n_substream_groups;
+               sg++) {
+            group_index = ParseAc4SgiSpecifier(reader);
+            max_group_index =
+                group_index > max_group_index ? group_index : max_group_index;
+          }
+          break;
+        default:
+          ParsePresentationConfigExtInfo(reader);
+          break;
+      }
+    }
+    RCHECK(reader->ReadBits(1, &ac4_presentation_v1_info.b_pre_virtualized));
+    RCHECK(reader->ReadBits(1, &ac4_presentation_v1_info.b_add_emdf_substreams));
+    ParseAc4PresentationSubstreamInfo(reader);
+  }
+  if (ac4_presentation_v1_info.b_add_emdf_substreams) {
+    RCHECK(reader->ReadBits(2, &ac4_presentation_v1_info.n_add_emdf_substreams));
+    if (ac4_presentation_v1_info.n_add_emdf_substreams == 0) {
+      ac4_presentation_v1_info.n_add_emdf_substreams =
+          ReadAc4VariableBits(reader, 2) + 4;
+    }
+    for (int i = 0; i < ac4_presentation_v1_info.n_add_emdf_substreams; i ++) {
+      ParseEmdfInfo(reader);
+    }
+  }
+  return true;
+}
+
+bool AC4Parser::ParseFrameRateMultiplyInfo(BitReader* reader) {
+  FrameRateMultiplyInfo frame_rate_multiply_info;
+  switch (ac4_toc.frame_rate_index) {
+    case 2:
+    case 3:
+    case 4:
+      RCHECK(reader->ReadBits(1, &frame_rate_multiply_info.b_multiplier));
+      if (frame_rate_multiply_info.b_multiplier) {  // b_multiplier
+        RCHECK(reader->ReadBits(
+            1, &frame_rate_multiply_info.multiplier_bit));  // multiplier_bit
+      }
+      break;
+    case 0:
+    case 1:
+    case 7:
+    case 8:
+    case 9:
+      RCHECK(reader->ReadBits(1, &frame_rate_multiply_info.b_multiplier));
+      break;
+    default:
+      break;
+  }
+  return true;
+}
+bool AC4Parser::ParseFrameRateFractionsInfo(BitReader* reader) {
+  FrameRateFractionsInfo frame_rate_fractions_info;
+  //int frame_rate_fraction = 1;
+  if (ac4_toc.frame_rate_index >= 5 && ac4_toc.frame_rate_index <= 9) {
+    RCHECK(
+        reader->ReadBits(1, &frame_rate_fractions_info.b_frame_rate_fraction));
+    // if (frame_rate_fractions_info.b_frame_rate_fraction == 1) {
+    //   frame_rate_fraction = 2;
+    // }
+  }
+  if (ac4_toc.frame_rate_index >= 10 && ac4_toc.frame_rate_index <= 12) {
+    RCHECK(
+        reader->ReadBits(1, &frame_rate_fractions_info.b_frame_rate_fraction));
+    if (frame_rate_fractions_info.b_frame_rate_fraction == 1) {
+      RCHECK(reader->ReadBits(
+          1, &frame_rate_fractions_info.b_frame_rate_fraction_is_4));
+      // if (frame_rate_fractions_info.b_frame_rate_fraction_is_4 == 1){
+      //   frame_rate_fraction = 4;
+      // } else {
+      //   frame_rate_fraction = 2;
+      // }
+      
+    }
+  }
+  return true;
+}
+bool AC4Parser::ParseEmdfInfo(BitReader* reader) {
```

Comment:
style nit: blank lines between methods please

more style: this method is hard to read with no blank lines separating the parsing into logical paragraphs